File: infrastructures/ddd/typescript/src/infrastructure/repositories/IExampleQueryRepository.ts
import { IQueryRepository } from './infra/IQueryRepository';
import { ExampleReadEntity } from '../persistence/read/ExampleReadEntity';

/**
 * Query repository interface for Example read entities.
 * Repository interfaces:
 * - Live in /infrastructure/repositories/
 * - Start with I prefix
 * - Extend IQueryRepository for read operations
 * - Implementations go in /infrastructure/repositories/impl/
 * 
 * Query repositories:
 * - Return ReadEntities directly to services
 * - No BMO mapping in read path for performance
 * - Optimized for query operations
 */
export interface IExampleQueryRepository extends IQueryRepository<ExampleReadEntity, string> {
  /**
   * Lists active examples with pagination.
   * Returns only active examples.
   */
  ListActiveAsync(p_skip: number, p_take: number): Promise<ExampleReadEntity[]>;

  /**
   * Lists examples by owner with pagination.
   */
  ListByOwnerAsync(p_ownerId: string, p_skip: number, p_take: number): Promise<ExampleReadEntity[]>;
}

File: infrastructures/ddd/python/src/infrastructure/repositories/i_example_query_repository.py
"""
Query repository interface for Example domain.
Repository interfaces:
- Live in /infrastructure/repositories/
- Start with I prefix
- Implementations go in /infrastructure/repositories/impl/
"""

from infrastructure.repositories.infra import IQueryRepository
from infrastructure.persistence.read.example_read_entity import ExampleReadEntity


class IExampleQueryRepository(IQueryRepository[ExampleReadEntity, str]):
    """
    Query repository interface for Example domain.
    
    This interface extends the base query repository with
    Example-specific query operations if needed.
    """
    pass
